Tasks/Common/VstsAzureHelpers_/InitializeFunctions.ps1

```diff
@@ -251,84 +251,89 @@ function Initialize-AzureSubscription {
 }
 
 
-# Get the Bearer Access Token from the Endpoint
+# Get the Bearer Access Token for Managed Identity Authentication scheme
 function Get-MsiAccessToken {
     [CmdletBinding()]
-    param([Parameter(Mandatory=$true)] $endpoint,
-        [Parameter(Mandatory=$true)] $retryCount,
-        [Parameter(Mandatory=$true)] $timeToWait)
+    param(
+        [Parameter(Mandatory=$true)] $endpoint,
+        [int] $retryLimit = 5,
+        [int] $timeToWait = 2000
+    )
+
+    Write-Verbose "Fetching access token for Managed Identity authentication from Azure Instance Metadata Service."
 
-    $msiClientId = "";
-    if($endpoint.Data.msiClientId){
-        $msiClientId  =  "&client_id=" + $endpoint.Data.msiClientId;
+    if ($endpoint.Data.msiClientId) {
+        $msiClientIdQueryParameter = "&client_id=$($endpoint.Data.msiClientId)"
     }
-    $tenantId = $endpoint.Auth.Parameters.TenantId
 
-    # Prepare contents for GET
-    $method = "GET"
-    $apiVersion = "2018-02-01";
-    $authUri = "http://169.254.169.254/metadata/identity/oauth2/token?api-version=" + $apiVersion + "&resource=" + $endpoint.Url + $msiClientId;
-    
-    # Call Rest API to fetch AccessToken
-    Write-Verbose "Fetching Access Token For MSI"
-    
-    try
-    {
-        $retryLimit = 5;
-        $proxyUri = Get-ProxyUri $authUri
-        if ($proxyUri -eq $null)
-        {
-            Write-Verbose "No proxy settings"
-            $response = Invoke-WebRequest -Uri $authUri -Method $method -Headers @{Metadata="true"} -UseBasicParsing
-        }
-        else
-        {
-            Write-Verbose "Using Proxy settings"
-            $response = Invoke-WebRequest -Uri $authUri -Method $method -Headers @{Metadata="true"} -UseDefaultCredentials -Proxy $proxyUri -ProxyUseDefaultCredentials -UseBasicParsing
-        }
+    $requestUri = "http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=$($endpoint.Url)$msiClientIdQueryParameter"
+    $requestHeaders = @{
+        Metadata = "true"
+    }
 
-        # Action on the based of response 
-        if(($response.StatusCode -eq 429) -or ($response.StatusCode -eq 500))
-        {
-            if($retryCount -lt $retryLimit)
-            {
-                $retryCount += 1
-                $waitedTime = 2000 + $timeToWait * 2
-                Start-Sleep -m $waitedTime
-                Get-MsiAccessToken $endpoint $retryCount  $waitedTime
+    $trialCount = 1
+    $retryableStatusCodes = @(429, 500, 502, 503, 504)
```

Comment:
Are these error codes same as what we retry for Azure Webapp?

Reply:
suggestion: Can we have this as separate Web client function that takes the URL and returns the response?
So that we can use this for other calls if required.

Reply:
add 409 as well

Reply:
Added 409.